broker/value_added/value_added.py

```diff
@@ -0,0 +1,211 @@
+
+    Returns:
+        light_curves (list): list of tuples formatted as required for input to RAPID.
+
+        cx_data      (list): list of dicts of candidate and xmatch info as needed for
+                             upload to BigQuery classification table.
+
+    """
+    # throw an error if received non-ZTF data
+    assert (survey=='ZTF'), "\nvalue_added.format_for_rapid() requires survey=='ZTF'"
+
+    light_curves = []
+    cx_dicts = {} # collect candidate and xmatch data to merge with RAPID results
+    oid_map = map_objectId_list(xmatch_list) # maps objectId's to list positions
+
+    for alert in alert_list:
+        oid = alert['objectId']
+        cid = alert['candidate']['candid']
+        cand_mjd = jd_to_mjd(alert['candidate']['jd'])
+        ra = alert['candidate']['ra']
+        dec = alert['candidate']['dec']
+
+        # Observation epoch data
+        fid_dict = {1:'g', 2:'r', 3:'i'}
+        mjd, flux, fluxerr, passband, photflag = ([] for i in range(5))
+
+        epochs = alert['prv_candidates'] + [alert['candidate']]
+        # if one epoch is missing a zeropoint, they should all be missing
+        zp_fallback, zp_in_keys = 26.0, [0 for i in range(len(epochs))]
+        for n, epoch in enumerate(epochs):
+            try:
+                assert epoch['magpsf'] is not None # magpsf is null for nondetections
+                # test this. try setting magnitude to epoch['diffmaglim'] instead of skipping.
+            except:
+                continue # move to next epoch
+
+            # early schema(s) did not contain a magnitude zeropoint
+            if 'magzpsci' not in epoch.keys(): # fix this. do something better.
+                _warn('Epoch does not have zeropoint data. Setting to {}'.format(zp_fallback))
+                zp_in_keys[n] = 1
+                epoch['magzpsci'] = zp_fallback
+
+            mjd.append(jd_to_mjd(epoch['jd']))
+            f, ferr = mag_to_flux(epoch['magpsf'],epoch['magzpsci'],epoch['sigmapsf'])
+            flux.append(f)
+            fluxerr.append(ferr)
+            passband.append(fid_dict[epoch['fid']])
+            photflag.append(4096)  # fix this, determines trigger time
+                                    # (1st mjd where this == 6144)
+
+        # check that either all or no epochs with detections have missing zeropoint
+        if sum(zp_in_keys) not in [0, len(mjd)]:
+            err_msg = ("Inconsistent zeropoint values in the epochs of alert {}."
+                        "Cannot continue with classification.").format(oid)
+            assert False, err_msg
+
+        # Set trigger date. fix this.
+        photflag[np.where(flux==np.max(flux))[0][0]] = 6144
+
+        # skip classification if don't have g passband
+        if 'g' not in passband: continue
+        # if 'r' not in passband: continue # this hasn't been a problem yet
+
+        # MW dust extinction
+        # fix this. ZTF docs say ra, dec are in J2000 [deg]
```

Comment:
I would nest this code under `if 'g' in passband` for readability purposes. Also, I'm not fully clear why not having g is a problem.

Reply:
I'm not sure about the logic of nesting it under `if g in passband`. I'll ask you in person.

I'm actually not sure why g is a problem either, but RAPID rejects them. There are a few things it rejects that I need to dig into more.